### PR TITLE
Improvements to HMD flying

### PR DESCRIPTION
**Pull request by hyperlogic**
In general, this PR should make flying while wearing an HMD more predictable and less nauseating.

* Embiggen the stick dead-spots for oculus touch controllers. This effectively splits the controller into directional zones. This should make driving/flying navigation more predictable and less prone to drifting in unintentional directions, which could induce nausea.
* Tighten action motor timescale while flying. This should cause less drift as the user changes direction and should lead to more comfort and control.
* Hmd head roll is only enabled if you are forward or backward faster than 2 meters per second. This should prevent un-intentional turning/rolling when hovering or moving at slow speeds.
* Documented hmdRoll tuning parameters accessible from JavaScript.
* Removed roll feathering code, because it is non-linear and can perhaps induce nausea.
* Tuned default dead spot and turning speeds to match Eagle Flight's defaults.

### CODE REVIEW

**Comment by hyperlogic** [comment]
https://highfidelity.fogbugz.com/f/cases/6482/Avatar-doesn-t-fly-straight-and-tends-to-veer-off-intended-course

**Comment by hyperlogic** [comment]
QA test plan in progress...

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5234/HighFidelity-Beta-PR11894-a4d70e89df86953545816c25440990130dba7136.dmg

**Comment by hyperlogic** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5237/HighFidelity-Beta-PR11894-a4d70e89df86953545816c25440990130dba7136.dmg

**Comment by hyperlogic** [comment]
npm keeps failing on windows build, build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5238/HighFidelity-Beta-PR11894-a4d70e89df86953545816c25440990130dba7136.dmg

**Comment by ZappoMan** [comment]
how does this PR handle this issue from the original bug report: "This is happening in desktop as well as HMD. When using desktop (forward only) we are able to rule out the possibility of the user not pushing it exactly forward when using hand controllers. "

**Comment by hyperlogic** [comment]
This PR only addresses flying in HMD mode, based on flying feedback from the greeter meeting. on Nov 20th.  At the moment I don't have any leads or a repro about the desktop issue, I have not been able to reproduce it, only the HMD issue.

Should I split this into two tickets?

**Comment by marko8904** [comment]
build this please

**Comment by Nex-Pro** [comment]
Awaiting Win build.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5244/HighFidelity-Beta-PR11894-a4d70e89df86953545816c25440990130dba7136.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5244/HighFidelity-Beta-PR11894-a4d70e89df86953545816c25440990130dba7136.dmg

**Comment by Nex-Pro** [comment]
In progress

**Comment by Nex-Pro** [comment]
### PASS

https://highfidelity.testrail.net/index.php?/tests/view/18235&group_by=cases:section_id&group_order=asc&group_id=1036

File: interface/src/avatar/MyAvatar.cpp [hunk 1]
```diff
@@ -1956,27 +1967,33 @@ void MyAvatar::updateOrientation(float deltaTime) {
...
-            } else {
-                speedFactor = glm::min(speed / _lastDrivenSpeed, 1.0f);
-            }
+        const float MIN_CONTROL_SPEED = 2.0f;  // meters / sec
+        const glm::vec3 characterForward = getWorldOrientation() * Vectors::UNIT_NEG_Z;
+        float forwardSpeed = glm::dot(characterForward, getWorldVelocity());
 
-            float direction = glm::dot(getWorldVelocity(), getWorldOrientation() * Vectors::UNIT_NEG_Z) > 0.0f ? 1.0f : -1.0f;
+        // only enable roll-turns if we are moving forward or backward at greater then MIN_CONTROL_SPEED
+        if (fabsf(forwardSpeed) >= MIN_CONTROL_SPEED) {
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyAvatar.cpp]: Someday we should make this threshold scale with avatar size.

File: interface/src/avatar/MyAvatar.cpp [hunk 2]
```diff
@@ -1956,27 +1967,33 @@ void MyAvatar::updateOrientation(float deltaTime) {
...
+
+            if (rollAngle > MIN_ROLL_ANGLE) {
+                // rate of turning is linearly proportional to rolAngle
+                rollAngle = glm::clamp(rollAngle, MIN_ROLL_ANGLE, MAX_ROLL_ANGLE);
+
+                // scale rollAngle into a value from zero to one.
+                float t = (rollAngle - MIN_ROLL_ANGLE) / (MAX_ROLL_ANGLE - MIN_ROLL_ANGLE);
+
+                float angularSpeed = rollSign * t * _hmdRollControlRate;
+                totalBodyYaw += direction * angularSpeed * deltaTime;
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyAvatar.cpp]: The coding standard asks for descriptive names, so the variable name `t` isn't descriptive enough here.  I'd accept it if it were for `time` but it appears to be a renormalized value of `rollAngle` and ultimately acts as a multiplicative factor on the final speed.  Maybe call it `rollFactor` or something?

File: interface/src/avatar/MyAvatar.cpp [hunk 3]
```diff
@@ -1956,27 +1967,33 @@ void MyAvatar::updateOrientation(float deltaTime) {
...
             float rollSign = rollAngle < 0.0f ? -1.0f : 1.0f;
             rollAngle = fabsf(rollAngle);
-            rollAngle = rollAngle > _hmdRollControlDeadZone ? rollSign * (rollAngle - _hmdRollControlDeadZone) : 0.0f;
 
-            totalBodyYaw += speedFactor * direction * rollAngle * deltaTime * _hmdRollControlRate;
+            const float MIN_ROLL_ANGLE = _hmdRollControlDeadZone;
+            const float MAX_ROLL_ANGLE = 90.0f;  // degrees
+
+            if (rollAngle > MIN_ROLL_ANGLE) {
+                // rate of turning is linearly proportional to rolAngle
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyAvatar.cpp]: rolAngle --> rollAngle

File: interface/src/avatar/MyAvatar.h [hunk 4]
```diff
@@ -737,12 +741,12 @@ private slots:
     bool _clearOverlayWhenMoving { true };
     QString _dominantHand { DOMINANT_RIGHT_HAND };
 
-    const float ROLL_CONTROL_DEAD_ZONE_DEFAULT = 8.0f; // deg
-    const float ROLL_CONTROL_RATE_DEFAULT = 2.5f; // deg/sec/deg
+    const float ROLL_CONTROL_DEAD_ZONE_DEFAULT = 8.0; // degrees
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyAvatar.h]: `8.0` --> `8.0f`

**Comment by marko8904** [comment]
build this please

**Comment by RyanDowne** [comment]
I really like the word Embiggen.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5487/HighFidelity-Beta-PR11894-47b52238d91bab462be296e85971bd0bb0229fea.dmg

**Comment by marko8904** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5494/HighFidelity-Beta-PR11894-47b52238d91bab462be296e85971bd0bb0229fea.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5494/HighFidelity-Beta-PR11894-47b52238d91bab462be296e85971bd0bb0229fea.dmg

**Pull request merged**